File: internal/crdt/identifier_test.go
package crdt

import "testing"

func TestIdentifierLess_ByCounter(t *testing.T) {
	a := Identifier{ClientID: 5, Counter: 1}
	b := Identifier{ClientID: 1, Counter: 2}

	if !a.Less(b) {
		t.Errorf("expected %+v < %+v (lower counter wins regardless of ClientID)", a, b)
	}
	if b.Less(a) {
		t.Errorf("expected %+v NOT < %+v", b, a)
	}
}

func TestIdentifierLess_TieBrokenByClientID(t *testing.T) {
	a := Identifier{ClientID: 1, Counter: 7}
	b := Identifier{ClientID: 2, Counter: 7}

	if !a.Less(b) {
		t.Errorf("expected %+v < %+v (same counter, lower ClientID wins)", a, b)
	}
	if b.Less(a) {
		t.Errorf("expected %+v NOT < %+v", b, a)
	}
}

func TestIdentifierLess_Equal(t *testing.T) {
	a := Identifier{ClientID: 3, Counter: 9}
	b := Identifier{ClientID: 3, Counter: 9}

	if a.Less(b) || b.Less(a) {
		t.Errorf("equal identifiers must not be Less than each other: %+v, %+v", a, b)
	}
}

func TestIdentifierGreater_IsInverseOfLess(t *testing.T) {
	a := Identifier{ClientID: 1, Counter: 1}
	b := Identifier{ClientID: 1, Counter: 2}

	if !b.Greater(a) {
		t.Errorf("expected %+v > %+v", b, a)
	}
	if a.Greater(b) {
		t.Errorf("expected %+v NOT > %+v", a, b)
	}
}

func TestIdentifierOrdering_NeverDependsOnFieldOrderOfComparison(t *testing.T) {
	// Sanity check that Less/Greater are proper strict-order inverses for a handful of pairs
	pairs := []struct{ a, b Identifier }{
		{Identifier{1, 1}, Identifier{2, 1}},
		{Identifier{1, 5}, Identifier{1, 6}},
		{Identifier{9, 3}, Identifier{1, 4}},
	}
	for _, p := range pairs {
		if p.a.Less(p.b) == p.b.Less(p.a) {
			t.Errorf("Less must be asymmetric for distinct identifiers: %+v vs %+v", p.a, p.b)
		}
	}
}

func TestRootID_IsRoot(t *testing.T) {
	if !RootID.IsRoot() {
		t.Error("RootID.IsRoot() should be true")
	}
	other := Identifier{ClientID: 1, Counter: 1}
	if other.IsRoot() {
		t.Error("a non-sentinel identifier should not report IsRoot() == true")
	}
}

File: web/src/lib/crdt/identifier.test.ts
import { describe, expect, it } from "vitest";
import { RootID, isGreater, isLess, isRoot } from "./identifier";

describe("identifier ordering", () => {
  it("orders by counter first", () => {
    const a = { clientId: 5, counter: 1 };
    const b = { clientId: 1, counter: 2 };
    expect(isLess(a, b)).toBe(true);
    expect(isLess(b, a)).toBe(false);
  });

  it("breaks ties on clientId when counters are equal", () => {
    const a = { clientId: 1, counter: 7 };
    const b = { clientId: 2, counter: 7 };
    expect(isLess(a, b)).toBe(true);
    expect(isLess(b, a)).toBe(false);
  });

  it("treats equal identifiers as neither less than the other", () => {
    const a = { clientId: 3, counter: 9 };
    const b = { clientId: 3, counter: 9 };
    expect(isLess(a, b)).toBe(false);
    expect(isLess(b, a)).toBe(false);
  });

  it("isGreater is the inverse of isLess", () => {
    const a = { clientId: 1, counter: 1 };
    const b = { clientId: 1, counter: 2 };
    expect(isGreater(b, a)).toBe(true);
    expect(isGreater(a, b)).toBe(false);
  });

  it("is asymmetric for every pair tested", () => {
    const pairs: [{ clientId: number; counter: number }, { clientId: number; counter: number }][] = [
      [{ clientId: 1, counter: 1 }, { clientId: 2, counter: 1 }],
      [{ clientId: 1, counter: 5 }, { clientId: 1, counter: 6 }],
      [{ clientId: 9, counter: 3 }, { clientId: 1, counter: 4 }],
    ];
    for (const [a, b] of pairs) {
      expect(isLess(a, b)).not.toBe(isLess(b, a));
    }
  });
});

describe("RootID", () => {
  it("reports isRoot true for RootID itself", () => {
    expect(isRoot(RootID)).toBe(true);
  });

  it("reports isRoot false for a non-sentinel identifier", () => {
    expect(isRoot({ clientId: 1, counter: 1 })).toBe(false);
  });
});